### PR TITLE
feat: nodejs refactoring, windows support

### PR DESCRIPTION
**Pull request by guybedford**
This PR fixes the Windows support and adds Windows and Mac CI. In the process I discovered that file descriptors on windows overlap with our 1, 2, 3 descriptors we reserve and any others we reserve for preopens so had to refactor the entire preopen system to be an indirect host preopen numbering - that is we have our own descriptors, which then point to the underlying host descriptors.

In the process I added instancing to the filesystem API so that one can create other custom filesystems with custom preopens, similarly to the Node.js WASI implementation, and handled this with a shared IO implementation as well.

I then also added proper streaming read support using the Node.js fs calls.

Finally this also updates the http implementation to then use the new shared IO implementation.

Fixes https://github.com/bytecodealliance/jco/issues/170.

### CODE REVIEW

**Review by yoshuawuyts — COMMENTED** [review]
Okay, I've looked it over. Overall this looks good. But I do worry a little about some of the coding conventions, though since we're using `Prettier` I'm not super sure whether we should diverge?

File: test/codegen.js [hunk 1]
```diff
@@ -2,7 +2,7 @@ import { readFile } from 'node:fs/promises';
 import { exec, jcoPath } from './helpers.js';
 import { strictEqual } from 'node:assert';
 
-const eslintPath = 'node_modules/.bin/eslint';
+const eslintPath = `node_modules/eslint/bin/eslint.js`;
```
[the remark this answers is not in the document]
Comment (guybedford) [reply test/codegen.js:5]: Yes, this is a shell script which wasn't working in Windows, and now instead the execution points to the JS file directly and instead we invoke `node ${eslintPath}`.

File: test/fixtures/wit/preview2-shim.wit [hunk 2]
```diff
@@ -0,0 +1,5 @@
+world preview2-shim {
+  export wasi:cli/command
+  export wasi:cli/environment
+  export wasi:cl=
```
[the remark this answers is not in the document]
Comment (guybedford) [reply test/fixtures/wit/preview2-shim.wit:4]: This is definitely a typo, thank you!

File: packages/preview2-shim/lib/nodejs/filesystem.js [hunk 3]
```diff
@@ -1,78 +1,436 @@
...
+      fd = bestPreopenMatch[0];
+      subpath = subpath.slice(bestPreopenMatch[1]);
+      if (subpath[0] === '/')
+        subpath = subpath.slice(1);
+    }
+    if (subpath.startsWith('.'))
+      subpath = subpath.slice(subpath[1] === '/' ? 2 : 1);
+    const descriptor = this.getDescriptor(fd);
+    if (descriptor.hostPreopen)
+      return descriptor.hostPreopen + (descriptor.hostPreopen.endsWith('/') ? '' : '/') + subpath;
```
[the remark this answers is not in the document]
Comment (guybedford) [reply packages/preview2-shim/lib/nodejs/filesystem.js:114]: As discussed, happy to switch into a braced pattern.

File: packages/preview2-shim/lib/nodejs/cli.js [hunk 4]
```diff
@@ -1,19 +1,9 @@
...
+import { argv, env, cwd } from 'node:process';
 
-export function _setArgs (args) {
-  _args = args;
-}
-
-export function _setCwd (cwd) {
-  _cwd = cwd;
-}
+let _env = Object.entries(env), _args = argv, _cwd = cwd();
```
[the remark this answers is not in the document]
Comment (guybedford) [reply packages/preview2-shim/lib/nodejs/cli.js:3]: As discussed, let's treat this as a dedicated linter update PR.

File: packages/preview2-shim/lib/nodejs/filesystem.js [hunk 5]
```diff
@@ -1,78 +1,424 @@
...
+const isWindows = platform === 'win32';
+
+class ReadableFileStream {
+  constructor (hostFd, position) {
+    this.hostFd = hostFd;
+    this.position = position;
+  }
+  read (len) {
+    const buf = new Uint8Array(Number(len));
+    const bytesRead = readSync(this.hostFd, buf, this.position ? { position: this.position } : undefined);
```
Comment (guybedford) [inline packages/preview2-shim/lib/nodejs/filesystem.js:15]: We should not be using the internal position here, but track the position as an offset within the stream itself.

File: packages/preview2-shim/lib/nodejs/filesystem.js [hunk 6]
```diff
@@ -1,78 +1,424 @@
...
+      },
+
+      dropDirectoryEntryStream(sid) {
+        io.dropStream(sid);
+      },
+
+      metadataHash(fd) {
+        const stats = fs.types.stat(fd);
+        const upper = BigInt(stats.dataModificationTimestamp.seconds);
+        return { upper, lower: BigInt(0) };
```
Comment (guybedford) [inline packages/preview2-shim/lib/nodejs/filesystem.js:405]: This should include the inode as well.

**Review by pchickey — APPROVED** [review]

**Pull request merged**